### DISCUSSION TITLE
Support for native login experiences with OAuth 2.0 First-Party Applications

### DISCUSSION THREAD

**Discussion by embesozzi** [keycloak/keycloak#25014] [Ideas]
As an IAM Architect based on my experience, the user experience of opening a browser in a mobile application for the federation process is not optimal. Therefore, offering more control of the authentication process to native applications is going to enhance the login experience.
As a reference, we have the **DRAFT** OAuth 2.0 for First-Party Native Applications, which provides some high-level guidelines:
- https://datatracker.ietf.org/doc/draft-parecki-oauth-first-party-native-apps/00/

Other vendors such ForgeRock or now Microsoft Entra, are offering similar user experiences:
- https://backstage.forgerock.com/docs/sdks/latest/sdks/deployment/embeddedlogin.html
- https://devblogs.microsoft.com/identity/introducing-native-auth/ (beta)

ForgeRock implemented it in a custom way; Microsoft's implementation is currently unknown.

Concerns:
- We don't have an RFC (just a draft) to follow the guidelines, and it seems that each IAM vendor implemented it in a custom way.
- The OpenID Foundation, so far, is not involved.

**Comment by embesozzi** [comment]
Just a quick follow-up. Here is a demo showing the user experience with the proposed standard **OAuth 2.0 First-Party Applications**.

https://github.com/user-attachments/assets/69dcde11-92d2-4166-89f4-2a5ece6bfeea



A high-level overview of the required customizations I developed for the demo:
- A custom `RealmResourceProvider` called `AuthorizationChallengeEndpoint` that supports the proposed standard implementing the endpoint `authorization_challenge_endpoint`.
- A custom `OAuth2ChallengeErrorRepresentation` for managing the AuthorizationChallenge response.
- A custom authenticator called `PasskeysAutofill` that supports this new standard.
- Lastly, a customization of `keycloak-js` for handling the metadata returned by the AuthorizationChallengeEndpoint for the Passkeys Authenticator.

In the demo, Keycloak is configured with an extension and a new flow that supports passkeys using a new authenticator. Both users have already registered their passkeys in Keycloak.
The Bank demo portal is configured with the JS SDK implementing OAuth 2.0 for First-Party Applications with Keycloak. For the sake of the demo, I'm showing the login process and step-up authentication to demonstrate the cool UX. Everything is implemented by the app based on the information returned by the IdP, with no redirection required.

[1] https://datatracker.ietf.org/doc/draft-parecki-oauth-first-party-apps/

**Reply by codewithbisky** [reply]
I like the idea and it is easy to follow. Where do we get the extension? I want to implement this.

**Comment by dteleguin** [comment]
The document has been adopted as an Internet Draft: https://www.ietf.org/archive/id/draft-ietf-oauth-first-party-apps-00.html

**Comment by embesozzi** [comment]
See the related discussion on refactoring authenticators to support an API-based authentication approach https://github.com/keycloak/keycloak/discussions/36924

**Comment by AhmedAssaf** [comment]
@embesozzi , is there any updated regarding native OAuth 2 for mobile

**Reply by nickmccally** [reply]
curious as well
